File: quickbindadapter/src/main/java/com/xiaobin/quickbindadapter/BaseLoadView.java
package com.xiaobin.quickbindadapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;

public abstract class BaseLoadView<T extends ViewDataBinding> {

    enum LoadMoreState {
        LOADING_MORE, LOAD_SUCCESS, LOAD_FAIL, LOAD_COMPLETE
    }

    //加载更多
    private LoadMoreState loadMoreState = LoadMoreState.LOAD_SUCCESS;

    private T loadView;
    private BindHolder bindHolder;

    public BindHolder get(ViewGroup parent) {
        createView(parent);
        if (bindHolder != null)
            return bindHolder;
        else
            return null;
    }

    public LoadMoreState getLoadMoreState() {
        return loadMoreState;
    }

    protected abstract int getLayoutId();

    public void createView(ViewGroup parent) {
        loadView = DataBindingUtil.inflate(
                LayoutInflater.from(parent.getContext()),
                getLayoutId(),
                parent, false);
        bindHolder = new BindHolder(loadView.getRoot());
        initView(loadView);
    }

    public void isLoadMore() {
        if (loadMoreState == LoadMoreState.LOADING_MORE) return;
        loadMoreState = LoadMoreState.LOADING_MORE;
        onLoadMore(loadView);
    }

    public void isLoadMoreEnd() {
        if (loadMoreState == LoadMoreState.LOAD_COMPLETE) return;
        loadMoreState = LoadMoreState.LOAD_COMPLETE;
        onLoadEnd(loadView);
    }

    public void isLoadMoreSuccess() {
        if (loadMoreState == LoadMoreState.LOAD_SUCCESS) return;
        loadMoreState = LoadMoreState.LOAD_SUCCESS;
        onLoadSuccess(loadView);
    }

    public void isLoadMoreFail() {
        if (loadMoreState == LoadMoreState.LOAD_FAIL) return;
        loadMoreState = LoadMoreState.LOAD_FAIL;
        onLoadFail(loadView);
    }

    protected abstract void initView(T loadView);

    /**
     * 正在加载更多
     */
    protected abstract void onLoadMore(T loadView);

    /**
     * 加载完成，没有更多
     */
    protected abstract void onLoadEnd(T loadView);

    /**
     * 加载成功
     */
    protected abstract void onLoadSuccess(T loadView);

    /**
     * 加载失败
     */
    protected abstract void onLoadFail(T loadView);
}

File: quickbindadapter/src/main/java/com/xiaobin/quickbindadapter/BaseLoadView.kt
package com.xiaobin.quickbindadapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.lifecycle.LifecycleOwner

abstract class BaseLoadView<T : ViewDataBinding?>(private val layoutId: Int) {

    enum class LoadMoreState {
        LOADING, SUCCESS, FAILED, NO_MORE, WAIT_LOADING, NO_STATE
    }

    private var viewHolder: BindHolder? = null

    //加载更多
    var loadMoreState = LoadMoreState.NO_STATE
        private set

    private var loadView: T? = null

    fun createViewHolder(
        parent: ViewGroup,
        lifecycleOwner: LifecycleOwner? = null,
        onClickLoadMore: View.OnClickListener? = null
    ): BindHolder {
        loadView = DataBindingUtil.inflate(
            LayoutInflater.from(parent.context),
            layoutId,
            parent, false
        )
        initView(loadView)
        loadView!!.root.setOnClickListener {
            if (loadMoreState == LoadMoreState.NO_MORE ||
                loadMoreState == LoadMoreState.LOADING ||
                loadMoreState == LoadMoreState.SUCCESS ||
                loadView == null
            ) {
                //加载中，加载结束并且没有更多数据时，不允许触发点击事件
                return@setOnClickListener
            }
            onClickLoadMore?.onClick(it)
        }
        //恢复状态
        when (loadMoreState) {
            LoadMoreState.LOADING -> isLoading(true)
            LoadMoreState.SUCCESS -> isLoadMoreSuccess(true)
            LoadMoreState.FAILED -> isLoadMoreFailed(true)
            LoadMoreState.NO_MORE -> isNoMoreData(true)
            LoadMoreState.WAIT_LOADING -> isWaitLoading(true)
            LoadMoreState.NO_STATE -> {}
        }
        viewHolder = BindHolder(loadView!!, lifecycleOwner)
        return viewHolder!!
    }

    /**
     * 变更为 等待用户点击加载更多 状态
     * 禁止了自动加载更多的时候，才会调用这个方法
     */
    fun isWaitLoading(ignore: Boolean = false) {
        if (!ignore && (loadMoreState == LoadMoreState.WAIT_LOADING || loadView == null)) return
        loadMoreState = LoadMoreState.WAIT_LOADING
        onStateChange(loadView!!, LoadMoreState.WAIT_LOADING)
    }

    /**
     * 变更为 正在加载更多 状态
     */
    fun isLoading(ignore: Boolean = false) {
        if (!ignore && (loadMoreState == LoadMoreState.LOADING || loadView == null)) return
        loadMoreState = LoadMoreState.LOADING
        onStateChange(loadView!!, LoadMoreState.LOADING)
    }

    /**
     * 变更为 加载更多完成且无更多数据 状态
     */
    fun isNoMoreData(ignore: Boolean = false) {
        if (!ignore && (loadMoreState == LoadMoreState.NO_MORE || loadView == null)) return
        loadMoreState = LoadMoreState.NO_MORE
        onStateChange(loadView!!, LoadMoreState.NO_MORE)
    }

    /**
     * 变更为 加载更多成功 状态
     */
    fun isLoadMoreSuccess(ignore: Boolean = false) {
        if (!ignore && (loadMoreState == LoadMoreState.SUCCESS || loadView == null)) return
        loadMoreState = LoadMoreState.SUCCESS
        onStateChange(loadView!!, LoadMoreState.SUCCESS)
    }

    /**
     * 变更为 加载更多失败了 状态
     */
    fun isLoadMoreFailed(ignore: Boolean = false) {
        if (!ignore && (loadMoreState == LoadMoreState.FAILED || loadView == null)) return
        loadMoreState = LoadMoreState.FAILED
        onStateChange(loadView!!, LoadMoreState.FAILED)
    }

    protected abstract fun initView(loadView: T?)

    /**
     * 状态变化
     */
    protected abstract fun onStateChange(loadView: T, state: LoadMoreState)

}